### PR TITLE
root: fix @loader_path on macOS

### PR DESCRIPTION
**Pull request by aphecetche**
@drbenmorgan @gartung  @greenc-FNAL @HadrienG2 @marcmengel @vitodb @vvolkl

this is an updated version of #37020

which is solving the following issue : 

```
python -c 'import ROOT'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/laurent/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/root-6.32.00-q3ogmdakq2pdy2rsvav5ibciru3j43ux/lib/root/ROOT/__init__.py", line 25, in <module>
    import cppyy
  File "/Users/laurent/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/root-6.32.00-q3ogmdakq2pdy2rsvav5ibciru3j43ux/lib/root/cppyy/__init__.py", line 81, in <module>
    from ._cpython_cppyy import *
  File "/Users/laurent/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/root-6.32.00-q3ogmdakq2pdy2rsvav5ibciru3j43ux/lib/root/cppyy/_cpython_cppyy.py", line 21, in <module>
    c = loader.load_cpp_backend()
        ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/laurent/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/root-6.32.00-q3ogmdakq2pdy2rsvav5ibciru3j43ux/lib/root/cppyy_backend/loader.py", line 79, in load_cpp_backend
    raise RuntimeError("could not load cppyy_backend library")
RuntimeError: could not load cppyy_backend library
```

### CODE REVIEW

**Comment by wdconinc** [comment]
Since the patch context indicates that something similar is already done for Linux with `$ORIGIN` [1], do you think it makes sense to submit a pull request to ROOT so this can be fixed upstream? With an upstream pull request we can then add the patch by URL and not need it for the next release (or whenever it is merged).

I'm not sure I understand why this started failing now if rpath has been on by default since 6.25 and the @loader_path has been missing for multiple major releases now. What changed? For an upstream pull request, identifying that will certainly help in getting it merged promptly.

[1]: https://github.com/root-project/root/pull/8576/files#diff-dd81606cd750b320d9ed46d043eb5e5dceaf7e7dbcf2d4934cae69bbfc27138fR448

**Comment by aphecetche** [comment]
My original draft PR was with ROOT 6.28, so yes, the problem has been there since a while I would say.
I must re-test it but as far as I remember I've only observed the issue with a Spack installed root, most probably because Spack installs under `lib/root` (?)
But yes, I'll try an upstream fix.

**Comment by aphecetche** [comment]
actually, trying to reproduce it by compiling root "by hand" : 

```
cmake ../root -DCMAKE_GENERATOR=Ninja -DCMAKE_INSTALL_PREFIX=../install -Dxrootd=OFF
```

the libraries are installed directly under `../install/lib` (and thus in this case `python -c 'import ROOT'` does work fine) 

I now wonder what exactly in the Spack `root`'s `package.py` makes the output libraries go under `lib/root` instead of just under `lib` ? @wdconinc  any idea ?

**Comment by wdconinc** [comment]
I think it's the `gnuinstall` option, https://root.cern/install/build_from_source/#all-build-options. That forces gnu file hierarchy standard, which places libraries in lib/root.

**Comment by aphecetche** [comment]
gnuinstall it is indeed.

upstream PR https://github.com/root-project/root/pull/15925

File: var/spack/repos/builtin/packages/root/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/root/package.py]:
```suggestion
        patch(
            "https://github.com/root-project/root/pull/15925.diff?full_index=1",
            sha256="1937290a4d54cd2e3e8a8d23d93b8dedaca9ed8dcfdcfa2f0d16629ff53fb3b7",
            when="@6.28: +python"
        )
```
Should this apply starting with 6.26? Last change to the patch context is at https://github.com/root-project/root/commit/64e8de57cea981f856971dd9d934f4881deee771.

Comment (aphecetche) [reply var/spack/repos/builtin/packages/root/package.py]: will have to x-check when I find a moment but I believe root 6.26 is kind of too old to be compilable on my Mac, so I might not be able to test that actually...

Comment (wdconinc) [reply var/spack/repos/builtin/packages/root/package.py]: I see that the `root` package has:
```
    # See https://github.com/root-project/root/issues/11714
    if sys.platform == "darwin" and macos_version() >= Version("13"):
        conflicts("@:6.26.09", msg="macOS 13 support was added in root 6.26.10")
```
I don't know when macOS 12 reaches end of life, but it seems only 6.26.10 and 6.26.14 would be under consideration here.

If we we start the patch at 6.28, maybe we can alternatively just add a conflict with older version for any macos version.

File: var/spack/repos/builtin/packages/root/loader.patch
Comment (wdconinc) [inline var/spack/repos/builtin/packages/root/loader.patch]: The file var/spack/repos/builtin/packages/root/loader.patch isn't used anymore. Once removed we can move ahead and merge here.

File: var/spack/repos/builtin/packages/root/package.py
Comment (wdconinc) [reply var/spack/repos/builtin/packages/root/package.py]: Added the last comment here as conflict. Now, on macOS 12, the patch is either applied (`@6.28: +python`) or a conflict is thrown (`@:6.27 +python`).

**Review by wdconinc — APPROVED** [review]
I'll approve this but give the maintainers a few days to evaluate.

**Review by gartung — APPROVED** [review]

**Pull request merged**
